include/deal.II/lac/petsc_snes.h

```diff
@@ -0,0 +1,477 @@
+
+    /**
+     * Solve the nonlinear system of equations $F(x) = 0$.
+     *
+     * This function returns the number of iterations.
+     * The vector @p x must contain the initial guess.
+     * Upon returning, the @p x vector contains the solution.
+     *
+     * Here we also set the matrices to handle Jacobians.
+     */
+    unsigned int
+    solve(VectorType &x, AMatrixType &A, PMatrixType &P);
+
+    /**
+     * Solve the affine nonlinear system of equations $F(x) = b$.
+     *
+     * This function returns the number of iterations.
+     * The vector @p x must contain the initial guess.
+     * Upon returning, the @p x vector contains the solution.
+     *
+     * Upon returning, the @p x vector contains the solution.
+     */
+    unsigned int
+    solve(VectorType &x, const VectorType &b);
+
+    /**
+     * Solve the affine nonlinear system of equations $F(x) = b$.
+     *
+     * This function returns the number of iterations.
+     * The vector @p x must contain the initial guess.
+     * Upon returning, the @p x vector contains the solution.
+     *
+     * Here we also set the matrix to handle the Jacobian.
+     */
+    unsigned int
+    solve(VectorType &x, const VectorType &b, PMatrixType &P);
+
+    /**
+     * Solve the affine nonlinear system of equations $F(x) = b$.
+     *
+     * This function returns the number of iterations.
+     * The vector @p x must contain the initial guess.
+     * Upon returning, the @p x vector contains the solution.
+     *
+     * Here we also set the matrices to handle Jacobians.
+     */
+    unsigned int
+    solve(VectorType &x, const VectorType &b, AMatrixType &A, PMatrixType &P);
+
+    /**
+     * Callback for the computation of the nonlinear residual $F(x)$.
+     */
+    std::function<int(const VectorType &x, VectorType &res)> residual;
+
+    /**
+     * Callback for the computation of the Jacobian
+     * $\dfrac{\partial F}{\partial x}$
+     */
+    std::function<int(const VectorType &x, AMatrixType &A, PMatrixType &P)>
+      jacobian;
+
+    /**
+     * Callback for monitoring the solution process.
+     *
+     * This function is called by NonlinearSolver at the beginning
+     * of each time step.
+     */
+    std::function<
+      int(const VectorType &x, const unsigned int step_number, real_type fval)>
+      monitor;
+
+    /**
+     * Setup Jacobian callback without matrices.
+     *
+     * This callback gives full control to users to set up the tangent
+     * operator $\dfrac{\partial F}{\partial x}$
+     *
+     * Solvers must be provided via NonlinearSolver::solve_for_jacobian_system.
+     */
+    std::function<int(const VectorType &x)> setup_jacobian;
+
+    /**
+     * Solution of the Jacobian system set up with
+     * NonlinearSolver::setup_jacobian
+     *
+     * This is used as a preconditioner inside the Krylov process.
+     */
+    std::function<int(const VectorType &src, VectorType &dst)>
+      solve_for_jacobian_system;
```

Comment:
```suggestion
      solve_with_jacobian;
```
to make it similar to KINSOL/NOX-wrapper.

Reply:
So? Do we have a final agreement here? @bangerth @luca-heltai

Reply:
except for the fact that `solve_with_jacobian` takes an extra tolerance argument which is completely ignored here since in my design I don't put this burden on the users

Reply:
That's ok. I think consistency in naming is important, consistency in argument lists maybe not.